hare/broker.go

```diff
@@ -348,64 +353,71 @@ func (b *Broker) handleEarlyMessage(logger log.Log, layer types.LayerID, nodeID
 	return nil
 }
 
-func (b *Broker) cleanOldLayers() {
+func (b *Broker) CleanOldLayers(current types.LayerID) {
```

Comment:
Why did you decide to call this for each layer tick?

This is going to cause problems because the tracker can be deleted for a layer before the layer is unregistered, which can result in sending more than one malfeasance proof for a given identity.

Previously layers would get unregistered when the consensus process completed or at registration time if the limit was exceeded. It seems simplest to me to remove eligibility trackers at those points, which will keep eligibility trackers and outboxes in sync.

Reply:
> This is going to cause problems because the tracker can be deleted for a layer before the layer is unregistered

thanks. can you help me identity the codepath where this would happen?

`CleanOldLayers` figures out the highest layer without an entry in `outbox`, i.e. it's already unregistered, or never registered. when a node gets spammed, we will create a tracker for those early messages (same problem you spotted earlier with `pending`) and they will never get cleaned. it is safer to always clean up after each tick. because after `onTick` returns, the node either has an entry in outbox for the current layer or it doesnt.

i deliberately do not want to keep trackers and outbox in sync for 2 reasons:
- i don't want to get into the `isEarly` logic like `outbox`
- it gets clean up regardless of whether a layer register successfully or not.

Reply:
i also added `pending` to be cleaned up with `trackers`.

Reply:
> thanks. can you help me identity the codepath where this would happen?

Actually I think it's fine, I hadn't realised that CleanOldLayers is not going to affect currently running layers.